src/utils.cpp

```diff
@@ -304,6 +304,121 @@ std::vector<std::string> get_kernel_cflags(
   return cflags;
 }
 
+std::string get_cgroup_path_in_hierarchy(uint64_t cgroupid,
+                                         std::string base_path)
+{
+  static std::map<std::pair<uint64_t, std::string>, std::string> path_cache;
+  struct stat path_st;
+
+  auto cached_path = path_cache.find({ cgroupid, base_path });
+  if (cached_path != path_cache.end() &&
+      stat(cached_path->second.c_str(), &path_st) >= 0 &&
+      path_st.st_ino == cgroupid)
+    return cached_path->second;
+
+  // Check for root cgroup path separately, since recursive_directory_iterator
+  // does not iterate over base directory
+  if (stat(base_path.c_str(), &path_st) >= 0 && path_st.st_ino == cgroupid)
+  {
+    path_cache[{ cgroupid, base_path }] = "/";
+    return "/";
+  }
+
+  for (auto &path_iter :
+       std_filesystem::recursive_directory_iterator(base_path))
+  {
+    if (stat(path_iter.path().c_str(), &path_st) < 0)
+      return "";
+    if (path_st.st_ino == cgroupid)
+    {
+      // Base directory is not a part of cgroup path
+      path_cache[{ cgroupid, base_path }] = path_iter.path().string().substr(
+          base_path.length());
+      return path_cache[{ cgroupid, base_path }];
+    }
+  }
+
+  return "";
+}
+
+std::vector<std::pair<std::string, std::string>> get_cgroup_hierarchy_roots()
+{
+  // Get all cgroup mounts and their type (cgroup/cgroup2) from /proc/mounts
+  std::ifstream mounts_file("/proc/mounts");
+  std::vector<std::pair<std::string, std::string>> result;
+
+  const std::regex cgroup_mount_regex("(cgroup[2]?) (\\S*)[ ]?.*");
+  for (std::string line; std::getline(mounts_file, line);)
+  {
+    std::smatch match;
+    if (std::regex_match(line, match, cgroup_mount_regex))
+    {
+      result.push_back({ match[1].str(), match[2].str() });
+    }
+  }
+
+  mounts_file.close();
+  return result;
+}
+
+std::vector<std::pair<std::string, std::string>> get_cgroup_paths(
+    uint64_t cgroupid,
+    std::string filter)
+{
+  // TODO: Rewrite using std::views when C++20 support becomes common
+  auto roots = get_cgroup_hierarchy_roots();
```

Comment:
What's the expected size of this vector? There's a lot of copying below so it could be more efficient to rewrite it by a loop.

Reply:
On my system there are 18 cgroup hierarchy roots, which is not much, nevertheless I could improve the code.